src/wallet/gtest/test_wallet.cpp

```diff
@@ -1005,144 +1033,159 @@ TEST(WalletTests, SpentSproutNoteIsFromMe) {
-        ivk,
-        wtx.vShieldedOutput[0].ephemeralKey,
-        wtx.vShieldedOutput[0].cmu);
-    ASSERT_EQ(static_cast<bool>(maybe_pt), true);
-    auto maybe_note = maybe_pt.get().note(ivk);
-    ASSERT_EQ(static_cast<bool>(maybe_note), true);
-    auto note2 = maybe_note.get();
-
-    // Get witness to retrieve position of note B we want to spend
-    SaplingOutPoint sop0(wtx.GetHash(), 0);
-    auto spend_note_witness =  wtx.mapSaplingNoteData[sop0].witnesses.front();
-    auto maybe_nf = note2.nullifier(extfvk.fvk, spend_note_witness.position());
-    ASSERT_EQ(static_cast<bool>(maybe_nf), true);
-    auto nullifier2 = maybe_nf.get();
-
-    // NOTE: Not updating the anchor results in a core dump.  Shouldn't builder just return error?
-    // *** Error in `./zcash-gtest': double free or corruption (out): 0x00007ffd8755d990 ***
-    anchor = saplingTree.root();
-
-    // Create transaction to spend note B
-    auto builder2 = TransactionBuilder(consensusParams, 2);
-    builder2.AddSaplingSpend(expsk, note2, anchor, spend_note_witness);
-    builder2.AddSaplingOutput(extfvk.fvk.ovk, pk, 12500, {});
-    auto tx2 = builder2.Build().GetTxOrThrow();
-    EXPECT_EQ(tx2.vin.size(), 0);
-    EXPECT_EQ(tx2.vout.size(), 0);
-    EXPECT_EQ(tx2.vJoinSplit.size(), 0);
-    EXPECT_EQ(tx2.vShieldedSpend.size(), 1);
-    EXPECT_EQ(tx2.vShieldedOutput.size(), 2);
-    EXPECT_EQ(tx2.valueBalance, 10000);
-
-    CWalletTx wtx2 {&wallet, tx2};
-
-    // Fake-mine this tx into the next block
-    EXPECT_EQ(0, chainActive.Height());
-    CBlock block2;
-    block2.vtx.push_back(wtx2);
-    block2.hashMerkleRoot = block2.BuildMerkleTree();
-    block2.hashPrevBlock = blockHash;
-    auto blockHash2 = block2.GetHash();
-    CBlockIndex fakeIndex2 {block2};
-    mapBlockIndex.insert(std::make_pair(blockHash2, &fakeIndex2));
-    fakeIndex2.nHeight = 1;
-    chainActive.SetTip(&fakeIndex2);
-    EXPECT_TRUE(chainActive.Contains(&fakeIndex2));
-    EXPECT_EQ(1, chainActive.Height());
-
-    auto saplingNoteData2 = wallet.FindMySaplingNotes(wtx2).first;
-    ASSERT_TRUE(saplingNoteData2.size() > 0);
-    wtx2.SetSaplingNoteData(saplingNoteData2);
-    wtx2.SetMerkleBranch(block2);
-    wallet.AddToWallet(wtx2, true, NULL);
-
-    // Verify note B is spent. AddToWallet invokes AddToSpends which updates mapTxSaplingNullifiers
-    EXPECT_TRUE(wallet.IsSaplingSpent(nullifier2));
-
-    // Verify note B belongs to wallet.
-    EXPECT_TRUE(wallet.IsFromMe(wtx2));
-    ASSERT_TRUE(wallet.mapSaplingNullifiersToNotes.count(nullifier2));
+    SelectParams(CBaseChainParams::REGTEST);
+    int overwinterActivationHeight = 5;
+    int saplingActivationHeight = 30;
+    int canopyActivationHeight = 70;
+    UpdateNetworkUpgradeParameters(Consensus::UPGRADE_OVERWINTER, overwinterActivationHeight);
+    UpdateNetworkUpgradeParameters(Consensus::UPGRADE_SAPLING, saplingActivationHeight);
+    UpdateNetworkUpgradeParameters(Consensus::UPGRADE_CANOPY, canopyActivationHeight);
```

Comment:
This test is setting `chainActive` to a fake chain, that has a real height (of one or two blocks) which doesn't correspond to any of these heights. I don't think this will immediately cause an issue because we aren't calling `AddToWalletIfInvolvingMe` directly, but are instead emulating its behaviour here by calling `FindMySaplingNotes`. However, it's more difficult to reason about, and the correct way to handle this kind of activation pattern would be to make a valid fake chain.

Instead, remove these and have the `for` loop over the ZIP 212 cases call either `RegtestActivateSapling()` or `RegtestActivateCanopy()` (and the corresponding `RegtestDeactivate*` before the end of the loop).